src/common/libeventlog/eventlog.c

```diff
@@ -0,0 +1,331 @@
+
+    if (json_array_append_new (a, o) < 0) {
+        json_decref (o);
+        errno = ENOMEM;
+        return -1;
+    }
+
+    return 0;
+}
+
+json_t *eventlog_decode (const char *s)
+{
+    char *copy = NULL;
+    char *input;
+    json_t *a = NULL;
+    int save_errno;
+    char *ptr;
+
+    if (!s) {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    if (!(copy = strdup (s)))
+        goto enomem;
+    input = copy;
+
+    /* gotta have atleast 1 newline, if not empty string */
+    if (strlen (s) && !(ptr = strchr (input, '\n'))) {
+        errno = EINVAL;
+        goto error;
+    }
+
+    if (!(a = json_array ()))
+        goto enomem;
+
+    while ((ptr = strchr (input, '\n'))) {
+        (*ptr++) = '\0';
+
+        if (eventlog_entry_append (a, input) < 0)
+            goto error;
+
+        input = ptr;
+    }
+
+    free (copy);
+    return a;
+
+ enomem:
+    errno = ENOMEM;
+ error:
+    save_errno = errno;
+    free (copy);
+    json_decref (a);
+    errno = save_errno;
+    return NULL;
+}
+
+bool eventlog_entry_validate (json_t *entry)
+{
+    json_t *name;
+    json_t *timestamp;
+    json_t *context;
+
+    if (!json_is_object (entry)
+        || !(name = json_object_get (entry, "name"))
+        || !json_is_string (name)
+        || !(timestamp = json_object_get (entry, "timestamp"))
+        || !json_is_number (timestamp))
+        return false;
+
+    if ((context = json_object_get (entry, "context"))) {
+        if (!json_is_object (context))
+            return false;
+    }
+
+    return true;
+}
+
+static json_t *eventlog_entry_decode_wrapper (const char *entry,
+                                              bool trailing_newline)
+{
+    int len;
+    char *ptr;
+    json_t *o;
+
+    if (!entry)
+        goto einval;
+
+    if (!(len = strlen (entry)))
+        goto einval;
+
+    if (trailing_newline) {
+        if (entry[len - 1] != '\n')
+            goto einval;
+
+        ptr = strchr (entry, '\n');
+        if (ptr != &entry[len - 1])
+            goto einval;
+    }
+    else {
+        if ((ptr = strchr (entry, '\n')))
+            goto einval;
+    }
+
+    if (!(o = json_loads (entry, 0, NULL)))
+        goto einval;
+
+    if (!eventlog_entry_validate (o)) {
+        json_decref (o);
+        goto einval;
+    }
+
+    return o;
+
+ einval:
+    errno = EINVAL;
+    return NULL;
+}
+
+json_t *eventlog_entry_decode (const char *entry)
+{
+    return eventlog_entry_decode_wrapper (entry, true);
+}
+
+static int get_timestamp_now (double *timestamp)
+{
+    struct timespec ts;
+    if (clock_gettime (CLOCK_REALTIME, &ts) < 0)
+        return -1;
+    *timestamp = (1E-9 * ts.tv_nsec) + ts.tv_sec;
+    return 0;
+}
+
+static json_t *entry_build (double timestamp,
+                            const char *name,
+                            json_t *context)
+{
+    json_t *o = NULL;
+
+    if (timestamp < 0.
+        || !name
+        || !strlen (name)
+        || strchr (name, ' ') != NULL
```

Comment:
RFC 18 doesn't actually impose any restrictions on what is in the "name" string anymore, so prob OK to skip these checks for white space.  (note embedded newlines are escaped in JSON strings and cannot be confused with the event end of line)